httprule/httprule.go

```diff
@@ -35,28 +36,191 @@ var (
 	ErrInvalidHttpRule = fmt.Errorf("invalid HttpRule")
 )
 
-// ParseProtoResponse parses a proto message from a HTTPRule and an
-// io.Reader, typically an http.Response.Body. rule.ResponseBody may
-// only reference a top-level field of the response target proto
-// message. rule.ResponseBody may also be empty.
-func ParseProtoResponse(rule *pb.HttpRule, body io.Reader, target proto.Message) error {
+// ParseProtoResponse parses a http.Response using a HttpRule into a target
+// message. The HttpRule contains a specification of how the response body and
+// headers are mapped into the target proto message. The body JSON may map
+// directly to the target message, or it may map to a top-level field of the
+// target message. Response headers may reference any top-level scalar or
+// repeated scalar fields of the target message.
+//
+// The http.Response body is consumed but not closed.
+func ParseProtoResponse(rule *pb.HttpRule, resp *http.Response, target proto.Message) error {
 	if err := ValidateHTTPRule(rule); err != nil {
 		return err
 	}
 
-	b, err := io.ReadAll(body)
+	b, err := io.ReadAll(resp.Body)
 	if err != nil {
 		return fmt.Errorf("ParseProtoResponse: %w", err)
 	}
+	bodyTarget := target
 	if rule.ResponseBody != "" {
-		target, err = newField(rule.ResponseBody, target)
+		bodyTarget, err = newField(rule.ResponseBody, target)
 		if err != nil {
 			return err
 		}
 	}
 
-	if err := protojson.Unmarshal(b, target); err != nil {
-		return fmt.Errorf("cannot protojson unmarshal: %w", err)
+	if len(bytes.TrimSpace(b)) > 0 {
+		if err := protojson.Unmarshal(b, bodyTarget); err != nil {
+			return fmt.Errorf("cannot protojson unmarshal: %w", err)
+		}
+	}
```

Comment:
I wonder if all of this should go into a `ParseResponseBody` function now.
Could also have an early return for `if len(bytes.TrimSpace(b)) == 0 { ...`

Reply:
done